File: libraries/python/getpatter/utils/container_slot_manager.py
"""
ContainerSlotManager — in-process call slot gatekeeper for Cloudflare Container deployments.

Tracks active Telnyx WebSocket sessions by call_session_id, enforces MAX_CONTAINER_CALL_SLOTS,
and exposes a /capacity + /health HTTP endpoint polled by the Cloudflare Durable Object
edge router for routing decisions.
"""

from __future__ import annotations

import asyncio
import logging
import os
import resource
import socket
import sys
import time

from dataclasses import dataclass, asdict
from typing import Callable, Optional, Set

logger = logging.getLogger("getpatter")


@dataclass(frozen=True)
class CapacityStats:
    container_id: str
    status: str  # "HEALTHY" | "AT_CAPACITY"
    active_calls: int
    max_slots: int
    available_slots: int
    memory_rss_mb: int
    cpu_utilization_pct: float
    uptime_seconds: int


class ContainerSlotManager:
    def __init__(
        self,
        max_slots: Optional[int] = None,
        high_watermark_ratio: float = 0.80,
        http_port: Optional[int] = None,
        on_high_watermark: Optional[Callable[[int, int], None]] = None,
        cooldown_seconds: float = 120.0,
        on_cooldown_complete: Optional[Callable[[], None]] = None,
        container_id: Optional[str] = None,
    ) -> None:
        if max_slots is None:
            max_slots = int(os.environ.get("MAX_CONTAINER_CALL_SLOTS", "15"))
        self.max_slots = max_slots
        self.high_watermark_ratio = high_watermark_ratio
        self.cooldown_seconds = float(os.environ.get("CONTAINER_COOLDOWN_SECONDS", str(cooldown_seconds)))
        self.on_cooldown_complete = on_cooldown_complete
        self.container_id = container_id or os.environ.get("CONTAINER_ID") or socket.gethostname()
        self.on_high_watermark = on_high_watermark

        self._active_sessions: Set[str] = set()
        self._high_watermark_fired = False
        self._is_cooling_down = False
        self._cooldown_task: Optional[asyncio.TimerHandle] = None
        self._start_time = time.time()
        self._server: Optional[asyncio.AbstractServer] = None
        self._last_process_cpu = time.process_time()
        self._last_wall_cpu = time.time()
        self._cpu_percent = 0.0

        if http_port is None:
            port_env = os.environ.get("CAPACITY_HTTP_PORT", "8080")
            http_port = int(port_env) if port_env.isdigit() else 8080

        self._http_port = http_port

    async def start_server(self) -> None:
        if self._http_port > 0 and self._server is None:
            await self._start_http_server(self._http_port)

    def acquire(self, call_session_id: str) -> bool:
        """
        Attempt to acquire a call slot for the given call_session_id.
        Returns True if acquired (or re-acquired), False if container at capacity.
        """
        if call_session_id in self._active_sessions:
            return True
        if len(self._active_sessions) >= self.max_slots:
            logger.warning(
                f"[PATTER] ContainerSlotManager: at capacity ({len(self._active_sessions)}/{self.max_slots}), "
                f"rejecting {call_session_id}"
            )
            return False

        self._cancel_cooldown()
        self._active_sessions.add(call_session_id)
        self._maybe_fire_high_watermark()
        logger.debug(
            f"[PATTER] ContainerSlotManager: +slot {call_session_id} "
            f"({len(self._active_sessions)}/{self.max_slots})"
        )
        return True

    def release(self, call_session_id: str) -> None:
        """
        Release the slot held by call_session_id. Safe to call if id was never acquired.
        """
        if call_session_id not in self._active_sessions:
            return
        self._active_sessions.remove(call_session_id)
        if len(self._active_sessions) / self.max_slots < self.high_watermark_ratio:
            self._high_watermark_fired = False

        logger.debug(
            f"[PATTER] ContainerSlotManager: -slot {call_session_id} "
            f"({len(self._active_sessions)}/{self.max_slots})"
        )

        if len(self._active_sessions) == 0 and self.cooldown_seconds > 0:
            self._start_cooldown()

    @property
    def active_count(self) -> int:
        return len(self._active_sessions)

    @property
    def available_slots(self) -> int:
        return max(0, self.max_slots - len(self._active_sessions))

    @property
    def is_at_capacity(self) -> bool:
        return len(self._active_sessions) >= self.max_slots

    @property
    def is_cooling_down(self) -> bool:
        return self._is_cooling_down

    def _sample_cpu(self) -> float:
        now_wall = time.time()
        now_proc = time.process_time()
        wall_diff = now_wall - self._last_wall_cpu
        if wall_diff > 0:
            self._cpu_percent = ((now_proc - self._last_process_cpu) / wall_diff) * 100
        self._last_wall_cpu = now_wall
        self._last_process_cpu = now_proc
        return self._cpu_percent

    def get_capacity_stats(self) -> CapacityStats:
        usage = resource.getrusage(resource.RUSAGE_SELF)
        # rss in maxrss is kilobytes on Linux / bytes on macOS.
        if sys.platform == "darwin":
            rss_mb = int(usage.ru_maxrss / (1024 * 1024))
        else:
            rss_mb = int(usage.ru_maxrss / 1024)

        status = "HEALTHY"
        if self.is_at_capacity:
            status = "AT_CAPACITY"
        elif self._is_cooling_down:
            status = "DRAINING_COOLDOWN"

        return CapacityStats(
            container_id=self.container_id,
            status=status,
            active_calls=len(self._active_sessions),
            max_slots=self.max_slots,
            available_slots=self.available_slots,
            memory_rss_mb=rss_mb,
            cpu_utilization_pct=round(self._sample_cpu(), 1),
            uptime_seconds=int(time.time() - self._start_time),
        )

    def _start_cooldown(self) -> None:
        self._cancel_cooldown()
        self._is_cooling_down = True
        logger.info(
            f"[PATTER] ContainerSlotManager: 0 active calls — entering {self.cooldown_seconds}s cooldown "
            f"for background task/API polling completion"
        )
        try:
            loop = asyncio.get_running_loop()
            self._cooldown_task = loop.call_later(self.cooldown_seconds, self._on_cooldown_timer_expired)
        except RuntimeError:
            pass

    def _on_cooldown_timer_expired(self) -> None:
        self._is_cooling_down = False
        self._cooldown_task = None
        logger.info("[PATTER] ContainerSlotManager: cooldown period completed")
        if self.on_cooldown_complete:
            self.on_cooldown_complete()

    def _cancel_cooldown(self) -> None:
        if self._cooldown_task:
            self._cooldown_task.cancel()
            self._cooldown_task = None
        if self._is_cooling_down:
            self._is_cooling_down = False
            logger.info("[PATTER] ContainerSlotManager: cooldown interrupted by new incoming call")

    async def close(self) -> None:
        if self._server:
            self._server.close()
            await self._server.wait_closed()
            self._server = None

    def _maybe_fire_high_watermark(self) -> None:
        if self._high_watermark_fired:
            return
        ratio = len(self._active_sessions) / self.max_slots
        if ratio >= self.high_watermark_ratio:
            self._high_watermark_fired = True
            logger.info(
                f"[PATTER] ContainerSlotManager: high-watermark reached "
                f"({len(self._active_sessions)}/{self.max_slots}, {round(ratio * 100)}%) — signalling pre-warm"
            )
            if self.on_high_watermark:
                self.on_high_watermark(len(self._active_sessions), self.max_slots)

    async def _start_http_server(self, port: int) -> None:
        async def handle_request(reader: asyncio.StreamReader, writer: asyncio.StreamWriter) -> None:
            try:
                data = await reader.read(2048)
                request_line = data.decode("utf-8", errors="ignore").split("\r\n")[0]
                parts = request_line.split(" ")
                method = parts[0] if len(parts) > 0 else ""
                path = parts[1].split("?")[0] if len(parts) > 1 else ""

                if method != "GET":
                    response = b"HTTP/1.1 405 Method Not Allowed\r\nAllow: GET\r\n\r\n"
                elif path in ("/capacity", "/health"):
                    import json

                    stats = asdict(self.get_capacity_stats())
                    body = json.dumps(stats).encode("utf-8")
                    headers = (
                        f"HTTP/1.1 200 OK\r\n"
                        f"Content-Type: application/json\r\n"
                        f"Content-Length: {len(body)}\r\n\r\n"
                    ).encode("utf-8")
                    response = headers + body
                else:
                    response = b"HTTP/1.1 404 Not Found\r\n\r\n"

                writer.write(response)
                await writer.drain()
            except Exception as err:
                logger.warning(f"[PATTER] ContainerSlotManager: HTTP handle error: {err}")
            finally:
                writer.close()
                await writer.wait_closed()

        try:
            self._server = await asyncio.start_server(handle_request, "0.0.0.0", port)
            logger.info(f"[PATTER] ContainerSlotManager: capacity endpoint listening on :{port}")
        except Exception as err:
            logger.warning(f"[PATTER] ContainerSlotManager: HTTP server error: {err}")


container_slot_manager = ContainerSlotManager()

File: libraries/typescript/src/utils/container-slot-manager.ts
/**
 * ContainerSlotManager — in-process WebSocket call slot gatekeeper for Cloudflare Container deployments.
 *
 * Tracks active Telnyx WebSocket sessions by call_session_id, enforces MAX_CONTAINER_CALL_SLOTS,
 * and exposes a /capacity + /health HTTP endpoint polled by the Cloudflare Durable Object
 * edge router for routing decisions.
 *
 * Design rules:
 *   1. acquire(id) is idempotent — re-acquiring an already-active id returns true.
 *   2. release(id) is a no-op if the id was never acquired.
 *   3. High-watermark (default 80 %) fires onHighWatermark() ONCE per surge, resetting
 *      when the next release() brings active count below the watermark.
 *   4. The HTTP server on port 8080 is optional (pass httpPort: 0 to disable).
 *
 * Usage:
 *   import { containerSlotManager } from 'getpatter';
 *   const ok = containerSlotManager.acquire(callSessionId);
 *   if (!ok) ws.terminate(); // let Telnyx re-route to another container
 *   // ... handle call ...
 *   containerSlotManager.release(callSessionId);
 */

import * as http from 'node:http';
import * as os from 'node:os';
import express from 'express';
import { WebSocketServer, WebSocket as WSWebSocket } from 'ws';
import { getLogger } from '../logger';

export interface ContainerSlotManagerOptions {
  /**
   * Hard cap on simultaneous call slots.
   * Default: MAX_CONTAINER_CALL_SLOTS env var, or 15 if unset.
   */
  readonly maxSlots?: number;
  /**
   * Ratio [0–1] above which onHighWatermark fires.
   * Default 0.80 (fires when activeCalls / maxSlots >= 0.80).
   */
  readonly highWatermarkRatio?: number;
  /**
   * Port for the /capacity and /health HTTP endpoints.
   * Default: CAPACITY_HTTP_PORT env var, or 8080.
   * Set to 0 to disable the HTTP server entirely.
   */
  readonly httpPort?: number;
  /**
   * Callback fired once when active call count crosses the high-watermark threshold.
   * Use this to asynchronously request a new container instance from the Durable Object
   * capacity registry before the hard cap is reached.
   */
  readonly onHighWatermark?: (activeCalls: number, maxSlots: number) => void;
  /**
   * Cooldown period in milliseconds after active call count drops to 0.
   * Allows last-minute background tasks, tool executions, and API polling to complete.
   * Default: CONTAINER_COOLDOWN_MS env var, or 120,000 ms (2 minutes).
   */
  readonly cooldownMs?: number;
  /**
   * Optional callback fired when the 2-minute cooldown period completes.
   */
  readonly onCooldownComplete?: () => void;
  /**
   * Identifier included in /capacity JSON responses.
   * Default: CONTAINER_ID env var, or os.hostname().
   */
  readonly containerId?: string;
  /**
   * Optional callback to delegate incoming Telnyx WebSocket streams directly to Patter StreamHandler.
   */
  readonly onConnection?: (ws: WSWebSocket, callSessionId: string) => void;
  /**
   * On-Demand Dynamic Push Callback for Cloudflare Load Balancer origin updates.
   * Fired immediately (< 5ms) when active call slot allocation changes.
   */
  readonly onCapacityChanged?: (activeCalls: number, maxSlots: number, isSaturated: boolean) => void;
}

export interface CapacityStats {
  readonly containerId: string;
  readonly status: 'HEALTHY' | 'AT_CAPACITY' | 'DRAINING_COOLDOWN';
  readonly activeCalls: number;
  readonly maxSlots: number;
  readonly availableSlots: number;
  readonly memoryRssMb: number;
  readonly cpuUtilizationPct: number;
  readonly uptimeSeconds: number;
  readonly isCoolingDown?: boolean;
}

export class ContainerSlotManager {
  private readonly activeSessions = new Map<string, true>();
  readonly maxSlots: number;
  private readonly highWatermarkRatio: number;
  readonly containerId: string;
  private readonly onHighWatermark?: (active: number, max: number) => void;
  private readonly cooldownMs: number;
  private readonly onCooldownComplete?: () => void;
  private readonly onConnection?: (ws: WSWebSocket, callSessionId: string) => void;
  private readonly onCapacityChanged?: (active: number, max: number, isSaturated: boolean) => void;
  private highWatermarkFired = false;
  private isCoolingDownState = false;
  private cooldownTimer: ReturnType<typeof setTimeout> | null = null;
  private httpServer: http.Server | null = null;
  private readonly startTime = Date.now();
  private cpuPercent = 0;
  private lastCpuUsage = process.cpuUsage();
  private lastCpuTime  = Date.now();
  private cpuSampleTimer: ReturnType<typeof setInterval> | null = null;

  constructor(opts: ContainerSlotManagerOptions = {}) {
    this.maxSlots = opts.maxSlots ?? parseInt(process.env['MAX_CONTAINER_CALL_SLOTS'] ?? '15', 10);
    this.highWatermarkRatio = opts.highWatermarkRatio ?? 0.80;
    this.cooldownMs = opts.cooldownMs ?? parseInt(process.env['CONTAINER_COOLDOWN_MS'] ?? '120000', 10);
    this.onCooldownComplete = opts.onCooldownComplete;
    this.onConnection = opts.onConnection;
    this.onCapacityChanged = opts.onCapacityChanged;
    this.containerId =
      opts.containerId ??
      process.env['CONTAINER_ID'] ??
      os.hostname();
    this.onHighWatermark = opts.onHighWatermark;

    const httpPort = opts.httpPort ?? parseInt(process.env['CAPACITY_HTTP_PORT'] ?? '8080', 10);
    if (httpPort > 0) {
      this.startHttpServer(httpPort);
    }

    // Sample CPU every 5 s; unref so it never blocks process exit
    this.cpuSampleTimer = setInterval(() => this.sampleCpu(), 5_000);
    this.cpuSampleTimer.unref?.();
  }

  /**
   * Attempt to acquire a call slot for the given call_session_id.
   * Cancels active cooldown if the container receives a call.
   */
  acquire(callSessionId: string): boolean {
    if (this.activeSessions.has(callSessionId)) return true;
    if (this.activeSessions.size >= this.maxSlots) {
      getLogger().warn(
        `[PATTER] ContainerSlotManager: at capacity (${this.activeSessions.size}/${this.maxSlots}), ` +
          `rejecting ${callSessionId}`
      );
      return false;
    }
    // Cancel cooldown if active
    this.cancelCooldown();

    this.activeSessions.set(callSessionId, true);
    this.maybeFireHighWatermark();
    this.onCapacityChanged?.(this.activeSessions.size, this.maxSlots, this.activeSessions.size >= this.maxSlots);
    getLogger().debug(
      `[PATTER] ContainerSlotManager: +slot ${callSessionId} ` +
        `(${this.activeSessions.size}/${this.maxSlots})`
    );
    return true;
  }

  /**
   * Release the slot held by callSessionId. Safe to call even if id was never acquired.
   * If active session count drops to 0, starts 2-minute cooldown timer to allow outstanding
   * LLM background tasks and API polling to complete.
   */
  release(callSessionId: string): void {
    if (!this.activeSessions.delete(callSessionId)) return;
    // Reset high-watermark latch once load drops below threshold
    if (this.activeSessions.size / this.maxSlots < this.highWatermarkRatio) {
      this.highWatermarkFired = false;
    }
    this.onCapacityChanged?.(this.activeSessions.size, this.maxSlots, this.activeSessions.size >= this.maxSlots);
    getLogger().debug(
      `[PATTER] ContainerSlotManager: -slot ${callSessionId} ` +
        `(${this.activeSessions.size}/${this.maxSlots})`
    );

    // If 0 active calls, start 2-minute cooldown timer
    if (this.activeSessions.size === 0 && this.cooldownMs > 0) {
      this.startCooldown();
    }
  }

  get activeCount(): number {
    return this.activeSessions.size;
  }

  get availableSlots(): number {
    return Math.max(0, this.maxSlots - this.activeSessions.size);
  }

  get isAtCapacity(): boolean {
    return this.activeSessions.size >= this.maxSlots;
  }

  get isCoolingDown(): boolean {
    return this.isCoolingDownState;
  }

  getCapacityStats(): CapacityStats {
    const rss = process.memoryUsage().rss;
    let status: 'HEALTHY' | 'AT_CAPACITY' | 'DRAINING_COOLDOWN' = 'HEALTHY';
    if (this.isAtCapacity) {
      status = 'AT_CAPACITY';
    } else if (this.isCoolingDownState) {
      status = 'DRAINING_COOLDOWN';
    }

    return {
      containerId:      this.containerId,
      status,
      activeCalls:      this.activeSessions.size,
      maxSlots:         this.maxSlots,
      availableSlots:   this.availableSlots,
      memoryRssMb:      Math.round(rss / 1024 / 1024),
      cpuUtilizationPct: Math.round(this.cpuPercent * 10) / 10,
      uptimeSeconds:    Math.round((Date.now() - this.startTime) / 1000),
      isCoolingDown:    this.isCoolingDownState,
    };
  }

  private startCooldown(): void {
    this.cancelCooldown();
    this.isCoolingDownState = true;
    getLogger().info(
      `[PATTER] ContainerSlotManager: 0 active calls — entering ${this.cooldownMs / 1000}s cooldown ` +
        `for background task/API polling completion`
    );
    this.cooldownTimer = setTimeout(() => {
      this.isCoolingDownState = false;
      this.cooldownTimer = null;
      getLogger().info(`[PATTER] ContainerSlotManager: cooldown period completed`);
      this.onCooldownComplete?.();
    }, this.cooldownMs);
    this.cooldownTimer.unref?.();
  }

  private cancelCooldown(): void {
    if (this.cooldownTimer) {
      clearTimeout(this.cooldownTimer);
      this.cooldownTimer = null;
    }
    if (this.isCoolingDownState) {
      this.isCoolingDownState = false;
      getLogger().info(`[PATTER] ContainerSlotManager: cooldown interrupted by new incoming call`);
    }
  }

  /** Gracefully shut down the capacity HTTP server and CPU sampling timer. */
  async close(): Promise<void> {
    this.cancelCooldown();
    if (this.cpuSampleTimer) {
      clearInterval(this.cpuSampleTimer);
      this.cpuSampleTimer = null;
    }
    if (this.httpServer) {
      await new Promise<void>((resolve) => this.httpServer!.close(() => resolve()));
      this.httpServer = null;
    }
  }

  private maybeFireHighWatermark(): void {
    if (this.highWatermarkFired) return;
    const ratio = this.activeSessions.size / this.maxSlots;
    if (ratio >= this.highWatermarkRatio) {
      this.highWatermarkFired = true;
      getLogger().info(
        `[PATTER] ContainerSlotManager: high-watermark reached ` +
          `(${this.activeSessions.size}/${this.maxSlots}, ` +
          `${Math.round(ratio * 100)}%) — signalling pre-warm`
      );
      this.onHighWatermark?.(this.activeSessions.size, this.maxSlots);
    }
  }

  private sampleCpu(): void {
    const now   = Date.now();
    const usage = process.cpuUsage(this.lastCpuUsage);
    const elapsed = now - this.lastCpuTime;
    if (elapsed > 0) {
      this.cpuPercent = ((usage.user + usage.system) / 1000 / elapsed) * 100;
    }
    this.lastCpuUsage = process.cpuUsage();
    this.lastCpuTime  = now;
  }

  private startHttpServer(port: number): void {
    const app = express();
    app.disable('x-powered-by');

    // GET /health — container health status
    app.get('/health', (_req, res) => {
      const stats = this.getCapacityStats();
      res.json({
        status: 'healthy',
        timestamp: new Date().toISOString(),
        capacity: stats,
      });
    });

    // GET /capacity — slot capacity stats for load balancing
    app.get('/capacity', (_req, res) => {
      res.json(this.getCapacityStats());
    });

    // GET /status — runtime statistics and uptime
    app.get('/status', (_req, res) => {
      res.json({
        uptimeSeconds: Math.floor(process.uptime()),
        memoryUsageMb: Math.round(process.memoryUsage().rss / (1024 * 1024)),
        capacity: this.getCapacityStats(),
      });
    });

    // Clean JSON 404 handler for unmatched routes
    app.use((req, res) => {
      res.status(404).json({ error: 'Route not found', path: req.path });
    });

    this.httpServer = http.createServer(app);

    const wss = new WebSocketServer({ noServer: true });

    this.httpServer.on('upgrade', (request, socket, head) => {
      const host = request.headers.host || 'localhost';
      const url = new URL(request.url ?? '', `http://${host}`);
      if (url.pathname === '/media') {
        const callSessionId = url.searchParams.get('call_session_id') || 
                             (request.headers['x-call-id'] as string) || 
                             `session-${Math.random().toString(36).substring(7)}`;

        if (!this.acquire(callSessionId)) {
          socket.write('HTTP/1.1 503 Service Unavailable\r\nContent-Type: text/plain\r\n\r\nContainer at capacity\r\n');
          socket.destroy();
          return;
        }

        wss.handleUpgrade(request, socket, head, (ws) => {
          (ws as WSWebSocket & { callSessionId?: string }).callSessionId = callSessionId;
          wss.emit('connection', ws, request);
        });
      } else {
        socket.destroy();
      }
    });

    wss.on('connection', (ws: WSWebSocket & { callSessionId?: string }) => {
      const sessionId = ws.callSessionId || `session-${Math.random().toString(36).substring(7)}`;
      getLogger().info(`[PATTER] Container media stream connected (sessionId=${sessionId})`);

      if (this.onConnection) {
        // Delegate WebSocket connection directly to Patter StreamHandler
        this.onConnection(ws, sessionId);
      } else {
        // Default stub echo handler
        ws.on('message', (data) => {
          ws.send(JSON.stringify({ event: 'media_frame_ack', size: data.toString().length }));
        });
      }

      ws.on('close', () => {
        if (ws.callSessionId) {
          getLogger().info(`[PATTER] Container media stream closed (sessionId=${ws.callSessionId})`);
          this.release(ws.callSessionId);
        }
      });
    });

    this.httpServer.listen(port, '0.0.0.0', () => {
      getLogger().info(
        `[PATTER] ContainerSlotManager: capacity endpoint listening on 0.0.0.0:${port}`
      );
    });

    this.httpServer.on('error', (err) => {
      getLogger().warn(
        `[PATTER] ContainerSlotManager: HTTP server error: ${err.message}`
      );
    });
  }
}

/**
 * Process-level singleton slot manager.
 * Auto-initialises on first import; reads MAX_CONTAINER_CALL_SLOTS from env.
 */
export const containerSlotManager = new ContainerSlotManager();
